tests/integration_tests/platforms.py

```diff
@@ -0,0 +1,232 @@
+                self.settings.EXISTING_INSTANCE_ID
+            )
+            return
+        image_id = self._get_image()
+        launch_args = {
+            'image_id': image_id,
+            'user_data': self.user_data,
+            'wait': self.wait,
+        }
+        if self.instance_type:
+            launch_args['instance_type'] = self.instance_type
+        launch_args.update(self.launch_kwargs)
+        self.instance = self.client.launch(**launch_args)
+        print('Launched instance: {}'.format(self.instance))
+
+    def destroy(self):
+        self.instance.delete()
+
+    def exec(self, command):
+        return self.instance.execute(command)
+
+    def pull_file(self, remote_file, local_file):
+        self.instance.pull_file(remote_file, local_file)
+
+    def push_file(self, local_path, remote_path):
+        self.instance.push_file(local_path, remote_path)
+
+    def pull_from_file(self, remote_path) -> str:
+        tmp_file = NamedTemporaryFile('r')
+        self.pull_file(remote_path, tmp_file.name)
+        with tmp_file as f:
+            contents = f.read()
+        return contents
+
+    def push_to_file(self, remote_path, contents: str):
+        with NamedTemporaryFile('w', delete=False) as tmp_file:
+            tmp_file.write(contents)
+        self.push_file(tmp_file.name, remote_path)
+        os.unlink(tmp_file.name)
+
+    def snapshot(self):
+        return self.client.snapshot(self.instance, clean=True)
+
+    def _install_new_cloud_init(self, remote_script):
+        self.exec(remote_script)
+        version = self.exec('cloud-init -v').split()[-1]
+        print('Installed cloud-init version: {}'.format(version))
+        self.instance.clean()
+        image_id = self.snapshot()
+        print('Created new image: {}'.format(image_id))
+        set_current_image(image_id)
+        # self.instance.restart(wait=True)
+
+    def install_proposed_image(self):
+        print('Installing proposed image')
+        remote_script = (
+            '{sudo} echo deb "http://archive.ubuntu.com/ubuntu '
+            '$(lsb_release -sc)-proposed main" | '
+            '{sudo} tee /etc/apt/sources.list.d/proposed.list\n'
+            '{sudo} apt-get update -q\n'
+            '{sudo} apt-get install -qy cloud-init'
+        ).format(sudo='sudo' if self.use_sudo else '')
+        self._install_new_cloud_init(remote_script)
+
+    def install_ppa(self, repo):
+        print('Installing PPA')
+        remote_script = (
+            '{sudo} add-apt-repository {repo} -y && '
+            '{sudo} apt-get update -q && '
+            '{sudo} apt-get install -qy cloud-init'
+        ).format(sudo='sudo' if self.use_sudo else '', repo=repo)
+        self._install_new_cloud_init(remote_script)
+
+    def install_deb(self):
+        print('Installing deb package')
+        deb_path = integration_settings.IMAGE_SOURCE
+        deb_name = os.path.basename(deb_path)
+        remote_path = '/tmp/{}'.format(deb_name)
```

Comment:
Can it run after we've SSH'ed? This could be a problem during an SRU, right?

Reply:
> Can it run after we've SSH'ed?

I believe it could (though I don't have a reproducing system to hand)

> This could be a problem during an SRU, right?

Potentially, but I think a `cloud-init status --wait` before doing anything else means that we're out of the danger window.